src/ucp/core/ucp_am.c

```diff
@@ -1784,6 +1784,19 @@ const ucp_request_send_proto_t ucp_am_reply_proto = {
 void ucp_am_proto_request_zcopy_reset(ucp_request_t *request)
 {
     ucs_assert(request->send.msg_proto.am.header.reg_desc != NULL);
+
+    if ((request->send.msg_proto.am.flags & UCP_AM_SEND_FLAG_COPY_HEADER) &&
+        !(request->flags & UCP_REQUEST_FLAG_USER_HEADER_COPIED) &&
+        (request->send.msg_proto.am.header.length != 0)) {
+        request->send.msg_proto.am.header.ptr = ucs_mpool_set_get_inline(
+                &request->send.ep->worker->am_mps,
+                request->send.msg_proto.am.header.length);
+        request->flags |= UCP_REQUEST_FLAG_USER_HEADER_COPIED;
+    }
```

Comment:
Reset of a request may be called when it's already on pending queue, so header must be copied before the request is enqueued since user header might be freed at this time.
IMO need to add an assertion here but copy header at very first trying to send the request.

Reply:
Correct me if I'm wrong, in Zcopy don't we copy the user header to UCX's registered buffer at the very first try of send?

Reply:
Right, I missed that it's copied from `reg_desc`